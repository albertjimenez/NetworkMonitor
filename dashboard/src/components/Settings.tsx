import React, {useEffect} from "react";
import {useNats} from "../hooks/useNats.tsx";
import useReadFromKV from "../hooks/useReadFromKV.tsx";
import {Button, Divider, Form, Input, message, Typography} from "antd";
import {sendToKV} from "../hooks/sendToKV.tsx";
import {Constants} from "../data/Constants.ts";

const {Title} = Typography;

interface FormData {
    refreshTime: string,
    hostInterval: string
}

const Settings: React.FC = () => {
    const nc = useNats();

    const kv_refresh_time = useReadFromKV(Constants.SYSTEM_MONITOR_REFRESH_TIME).value;
    const kv_host_scan_interval = useReadFromKV(Constants.HOST_SCAN_INTERVAL_TIME).value;

    const [form] = Form.useForm();

    const onFinish = async (values: FormData) => {
        if (nc != null) {
            form.setFieldsValue({
                refreshTime: values.refreshTime,
                hostInterval: values.hostInterval,
            });
            await sendToKV(nc, Constants.SYSTEM_MONITOR_REFRESH_TIME, values.refreshTime);
            await sendToKV(nc, Constants.HOST_SCAN_INTERVAL_TIME, values.hostInterval);
            await message.success("Values updated successfully!");
        }
    };
    useEffect(() => {
        form.setFieldsValue({
            refreshTime: kv_refresh_time,
            hostInterval: kv_host_scan_interval,
        });
    }, [form, nc, kv_refresh_time, kv_host_scan_interval]);
    return <>
        <Title level={2}>Settings</Title>
        <Divider/>
        <Title level={3}>System Monitor</Title>
        <Form
            layout={"vertical"}
            form={form}
            onFinish={onFinish}>
            <Form.Item
                label="System Monitor Refresh Time"
                name="refreshTime"
                rules={[{required: true, message: "Please input a refresh time!"}]}
            >
                <Input type="number" placeholder="Enter refresh time in seconds"/>
            </Form.Item>
            <Divider/>
            <Title level={3}>Host Scan</Title>
            <Form.Item
                label="Host Scan Interval"
                name="hostInterval"
                rules={[{required: true, message: "Please input a host scan interval!"}]}
            >
                <Input type="number" placeholder="Enter host scan interval in seconds"/>
            </Form.Item>

            <Form.Item>
                <Button type="primary" htmlType="submit">
                    Submit
                </Button>
            </Form.Item>
        </Form>
    </>
}

export default Settings;
