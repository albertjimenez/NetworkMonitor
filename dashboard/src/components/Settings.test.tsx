import {render, screen} from '@testing-library/react';
import {beforeAll, describe, expect, it, vi} from 'vitest';
import Settings from './Settings';
import '@testing-library/jest-dom';
import {BrowserRouter} from "react-router-dom";

beforeAll(() => {
    window.matchMedia = vi.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: vi.fn(), // Deprecated
        removeListener: vi.fn(), // Deprecated
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
        dispatchEvent: vi.fn(),
    }));
});
// Mock useNats
vi.mock('../hooks/useNats.tsx', () => ({
    useNats: () => ({mockNatsConnection: true}),  // Return a mock NATS object
}));

// Mock useReadFromKV
vi.mock('../hooks/useReadFromKV.tsx', () => ({
    default: (key: string) => ({
        value: key === 'SYSTEM_MONITOR_REFRESH_TIME' ? '30' : '60',  // Mock values for KV keys
    }),
}));

// Mock sendToKV directly within vi.mock
vi.mock('../hooks/sendToKV.tsx', () => ({
    sendToKV: vi.fn(async () => Promise.resolve()),  // Mock sendToKV function inside vi.mock
}));

describe('Settings component', () => {
    it('renders correctly and matches snapshot', () => {
        const settings = render(<BrowserRouter>
            <Settings/>
        </BrowserRouter>);
        expect(settings).toMatchSnapshot();
    });
    it('renders the settings form with prefilled values from KV store', () => {
        render(<Settings/>);

        // Check if the input fields are populated with mocked KV values
        const refreshTimeInput = screen.getByPlaceholderText(/enter refresh time/i);
        const hostIntervalInput = screen.getByPlaceholderText(/enter host scan interval/i);

        expect(refreshTimeInput).toHaveValue(60);  // Mocked value from useReadFromKV
        expect(hostIntervalInput).toHaveValue(60);  // Mocked value from useReadFromKV
    });
});

