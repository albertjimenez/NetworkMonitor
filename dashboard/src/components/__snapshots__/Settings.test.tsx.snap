// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Settings component > renders correctly and matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h2
        class="ant-typography css-dev-only-do-not-override-14i19y2"
      >
        Settings
      </h2>
      <div
        class="ant-divider css-dev-only-do-not-override-14i19y2 ant-divider-horizontal"
        role="separator"
      />
      <h3
        class="ant-typography css-dev-only-do-not-override-14i19y2"
      >
        System Monitor
      </h3>
      <form
        class="ant-form ant-form-vertical css-dev-only-do-not-override-14i19y2"
      >
        <div
          class="ant-form-item css-dev-only-do-not-override-14i19y2 ant-form-item-has-success"
        >
          <div
            class="ant-row ant-form-item-row css-dev-only-do-not-override-14i19y2"
          >
            <div
              class="ant-col ant-form-item-label css-dev-only-do-not-override-14i19y2"
            >
              <label
                class="ant-form-item-required"
                for="refreshTime"
                title="System Monitor Refresh Time"
              >
                System Monitor Refresh Time
              </label>
            </div>
            <div
              class="ant-col ant-form-item-control css-dev-only-do-not-override-14i19y2"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <input
                    aria-required="true"
                    class="ant-input css-dev-only-do-not-override-14i19y2 ant-input-outlined ant-input-status-success"
                    id="refreshTime"
                    placeholder="Enter refresh time in seconds"
                    type="number"
                    value="60"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-divider css-dev-only-do-not-override-14i19y2 ant-divider-horizontal"
          role="separator"
        />
        <h3
          class="ant-typography css-dev-only-do-not-override-14i19y2"
        >
          Host Scan
        </h3>
        <div
          class="ant-form-item css-dev-only-do-not-override-14i19y2 ant-form-item-has-success"
        >
          <div
            class="ant-row ant-form-item-row css-dev-only-do-not-override-14i19y2"
          >
            <div
              class="ant-col ant-form-item-label css-dev-only-do-not-override-14i19y2"
            >
              <label
                class="ant-form-item-required"
                for="hostInterval"
                title="Host Scan Interval"
              >
                Host Scan Interval
              </label>
            </div>
            <div
              class="ant-col ant-form-item-control css-dev-only-do-not-override-14i19y2"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <input
                    aria-required="true"
                    class="ant-input css-dev-only-do-not-override-14i19y2 ant-input-outlined ant-input-status-success"
                    id="hostInterval"
                    placeholder="Enter host scan interval in seconds"
                    type="number"
                    value="60"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-form-item css-dev-only-do-not-override-14i19y2"
        >
          <div
            class="ant-row ant-form-item-row css-dev-only-do-not-override-14i19y2"
          >
            <div
              class="ant-col ant-form-item-control css-dev-only-do-not-override-14i19y2"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <button
                    class="ant-btn css-dev-only-do-not-override-14i19y2 ant-btn-primary"
                    type="submit"
                  >
                    <span>
                      Submit
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <h2
      class="ant-typography css-dev-only-do-not-override-14i19y2"
    >
      Settings
    </h2>
    <div
      class="ant-divider css-dev-only-do-not-override-14i19y2 ant-divider-horizontal"
      role="separator"
    />
    <h3
      class="ant-typography css-dev-only-do-not-override-14i19y2"
    >
      System Monitor
    </h3>
    <form
      class="ant-form ant-form-vertical css-dev-only-do-not-override-14i19y2"
    >
      <div
        class="ant-form-item css-dev-only-do-not-override-14i19y2 ant-form-item-has-success"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-14i19y2"
        >
          <div
            class="ant-col ant-form-item-label css-dev-only-do-not-override-14i19y2"
          >
            <label
              class="ant-form-item-required"
              for="refreshTime"
              title="System Monitor Refresh Time"
            >
              System Monitor Refresh Time
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control css-dev-only-do-not-override-14i19y2"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  aria-required="true"
                  class="ant-input css-dev-only-do-not-override-14i19y2 ant-input-outlined ant-input-status-success"
                  id="refreshTime"
                  placeholder="Enter refresh time in seconds"
                  type="number"
                  value="60"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-divider css-dev-only-do-not-override-14i19y2 ant-divider-horizontal"
        role="separator"
      />
      <h3
        class="ant-typography css-dev-only-do-not-override-14i19y2"
      >
        Host Scan
      </h3>
      <div
        class="ant-form-item css-dev-only-do-not-override-14i19y2 ant-form-item-has-success"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-14i19y2"
        >
          <div
            class="ant-col ant-form-item-label css-dev-only-do-not-override-14i19y2"
          >
            <label
              class="ant-form-item-required"
              for="hostInterval"
              title="Host Scan Interval"
            >
              Host Scan Interval
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control css-dev-only-do-not-override-14i19y2"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  aria-required="true"
                  class="ant-input css-dev-only-do-not-override-14i19y2 ant-input-outlined ant-input-status-success"
                  id="hostInterval"
                  placeholder="Enter host scan interval in seconds"
                  type="number"
                  value="60"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-form-item css-dev-only-do-not-override-14i19y2"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-14i19y2"
        >
          <div
            class="ant-col ant-form-item-control css-dev-only-do-not-override-14i19y2"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <button
                  class="ant-btn css-dev-only-do-not-override-14i19y2 ant-btn-primary"
                  type="submit"
                >
                  <span>
                    Submit
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
